### ISSUE TITLE
Add support for variadic functions in the catalog and built in concat function

### ISSUE THREAD

**Issue by jkosh44** [cmu-db/noisepage#1176]
The database catalog currently doesn't support registering functions with a variable number of arguments, which needs to be added. We also currently don't support the CONCAT built in function which needs to be added. 

The reason I'm including them in the same issue is because CONCAT has a variable amount of arguments and needs the database catalog to support this. Also implementing variadic functions in the catalog without a function that utilizes it will make testing more difficult. Therefore it makes sense to do them together.

Some things to consider for this issue
- When looking up the proc oid of a function we use it's signature (method name, argument count and types). We have an index on the pg proc table based on the method name and arguments. For a function with a variable number of arguments this won't work. 
https://github.com/cmu-db/noisepage/blob/master/src/binder/bind_node_visitor.cpp#L665
https://github.com/cmu-db/noisepage/blob/master/src/catalog/database_catalog.cpp#L2970-L3011
- Below is the contents of `pg_catalog.pg_proc` in postgres for concat. The meaning of all those columns can be found here: https://www.postgresql.org/docs/9.3/catalog-pg-proc.html
```
 oid  | proname | pronamespace | proowner | prolang | procost | prorows | provariadic | prosupport | prokind | prosecdef | proleakproof | proisstrict | proretset | provolatile | proparallel | pronargs | pronargdefaults | prorettype | proargtypes | proallargtypes | proargmodes | proargnames | proargdefaults | protrftypes |   prosrc    | probin | proconfig | proacl 
------+---------+--------------+----------+---------+---------+---------+-------------+------------+---------+-----------+--------------+-------------+-----------+-------------+-------------+----------+-----------------+------------+-------------+----------------+-------------+-------------+----------------+-------------+-------------+--------+-----------+--------
 3058 | concat  |           11 |       10 |      12 |       1 |       0 |        2276 | -          | f       | f         | f            | f           | f         | s           | s           |        1 |               0 |         25 | 2276        | {2276}         | {v}         |             |                |             | text_concat |        |           | 

```
Below is the contents of `pg_type` in postgres from type 2276. Meanings of the columns can be found here: https://www.postgresql.org/docs/9.3/catalog-pg-type.html
```
 oid  | typname | typnamespace | typowner | typlen | typbyval | typtype | typcategory | typispreferred | typisdefined | typdelim | typrelid | typelem | typarray | typinput | typoutput | typreceive | typsend | typmodin | typmodout | typanalyze | typalign | typstorage | typnotnull | typbasetype | typtypmod | typndims | typcollation | typdefaultbin | typdefault | typacl 
------+---------+--------------+----------+--------+----------+---------+-------------+----------------+--------------+----------+----------+---------+----------+----------+-----------+------------+---------+----------+-----------+------------+----------+------------+------------+-------------+-----------+----------+--------------+---------------+------------+--------
 2276 | any     |           11 |       10 |      4 | t        | p       | P           | f              | t            | ,        |        0 |       0 |        0 | any_in   | any_out   | -          | -       | -        | -         | -          | i        | p          | f          |           0 |        -1 |        0 |            0 |               |            | 
```

**Comment by jkosh44** [comment]
On my local branch here: https://github.com/jkosh44/terrier/tree/add-builtin-concat I've added in all the functionality for CONCAT in the execution engine and semantic checking.

**Comment by lmwnshn** [comment]
Fixed by #1139.
